src/blocks/battery/apc_ups.rs

```diff
@@ -0,0 +1,196 @@
+                .await
+                .error("Failed to connect to socket")?,
+            interval: interval.timer(),
+        })
+    }
+
+    async fn get_status(&self) -> Result<HashMap<String, String>> {
+        self.write(b"status").await?;
+
+        let result = self.read_response().await?;
+
+        let mut status_data: HashMap<String, String> = HashMap::new();
+        for line in result.lines() {
+            let (key, value) = line.split_once(':').unwrap();
+            status_data.insert(String::from(key.trim()), String::from(value.trim()));
+        }
+
+        Ok(status_data)
+    }
+
+    async fn write(&self, msg: &[u8]) -> Result<()> {
+        let msg_len = msg.len();
+        if msg_len >= 1 << 16 {
+            return Err(Error::new(
+                "msg is too long, it must be less than 2^16 characters long",
+            ));
+        }
+
+        // Write how many bytes of data are going to be sent (only need the last two octets)
+        // Write the actual message
+        let msgs = [&msg_len.to_be_bytes()[6..], msg];
+
+        for msg in msgs {
+            loop {
+                // Wait for the socket to be writable
+                self.stream
+                    .writable()
+                    .await
+                    .error("Socket closed unexpectedly")?;
+
+                // Try to write data, this may still fail with `WouldBlock`
+                // if the readiness event is a false positive.
+                match self.stream.try_write(msg) {
+                    Ok(_) => {
+                        break;
+                    }
+                    Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                        continue;
+                    }
+                    Err(e) => {
+                        return Err(Error::new(e.to_string()));
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn read_exact(&self, n: usize) -> Result<Vec<u8>> {
+        loop {
+            // Wait for the socket to be readable
+            self.stream
+                .readable()
+                .await
+                .error("Socket closed unexpectedly")?;
+
+            let mut buf = vec![0_u8; n];
+
+            // Try to read data, this may still fail with `WouldBlock`
+            // if the readiness event is a false positive.
+            match self.stream.try_read(&mut buf) {
+                Ok(_) => return Ok(buf),
+                Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                    continue;
+                }
+                Err(e) => {
+                    return Err(Error::new(e.to_string()));
+                }
+            }
+        }
+    }
+
+    async fn read_response(&self) -> Result<String> {
+        let mut buf = String::new();
+        loop {
+            let read_info = self.read_exact(2).await?;
+            let read_size = ((read_info[0] as usize) << 8) + (read_info[1] as usize);
```

Comment:
There is also [`AsyncReadExt::read_u16()`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncReadExt.html#method.read_u16)